### PR TITLE
Add README

### LINKED ISSUE #3: README: good practices sharing your secrets with third-party actions
### Discussed in https://github.com/Nautilus-Cyberneering/github-actions-secrets/discussions/2

<div type='discussions-op-text'>

<sup>Originally posted by **da2ce7** May 27, 2022</sup>
Is this the repository where workflows are run without permissions? hmmm?</div>

Write the README.

### PR DESCRIPTION
**Pull request by josecelano**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+No, there is not.
+
+**Is there a way to completely disable access to secrets in a workflow?**
+
+As you can see in this [workflow example](./.github/workflows/disable-github-token.yml), even if you disable the permissions for the `GITHUB_TOKEN` you are still able to get the secrets because you get them from the context not using the `GITHUB_TOKEN`.
+
+**Are contexts stored on disk in the GitHub runners?**
+
+As far as we know, they are not. But something we miss from the documentation it's a better explanation of the lifecycle of the runners and how it is the communication between GitHub and runners. We assume that the context for jobs is used only at the Github's servers. The secrets are shared with the runner only before executing a job and they are not stored on the disk.
```
Comment (cgbosse) [inline README.md:222]: ... the documentation **""is""** a better explanation of ...

File: README.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  - [Solution 1: Using CODEOWNERS configuration](#solution-1-using-codeowners-configuration)
+  - [Solution 2: Using branch protection rules and environment secrets](#solution-2-using-branch-protection-rules-and-environment-secrets)
+  - [Solution 3: Use a different repository to run MegaLinter](#solution-3-use-a-different-repository-to-run-megalinter)
+  - [Conclusion](#conclusion)
+- [How GitHub Actions can get access to secrets](#how-github-actions-can-get-access-to-secrets)
+- [Other questions](#other-questions)
+- [Links](#links)
+- [Hot to contribute](#how-to-contribute)
+
+We are using [MegaLinter](https://github.com/megalinter/megalinter) in most of our projects. It is a very useful GitHub Action to check your code, not only for basic things like the code format, but it also has some packages to detect security problems. But we have been wondering if the MegaLinter itself could be a security problem. This is the default workflow configuration when you install it:
```
Comment (cgbosse) [inline README.md:16]: I would start an additiona paragraph like so:

...security problems. However, we have also been wondering if the MegaLinter itself could be a security problem. 

Now, start by looking at it's default workflow configuration when you install it:

File: README.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+```
+
+You can check it on any of the [workflow executions](https://github.com/Nautilus-Cyberneering/github-actions-secrets/actions/workflows/mega-linter.yml) in this repo.
+
+As you can see there is an environment variable `-e GITHUB_TOKEN`. MegaLinter has 97 packages at the moment. That means there are a lot of dependencies that have access to the `GITHUB_TOKEN`.
```
Comment (cgbosse) [inline README.md:35]: As you can **"",""** see there is an environment variable `-e GITHUB_TOKEN`.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+- CODEOWNERS only works on pull requests. And in that case, `contents: write` permission is not granted.
+- We are talking about branches created on the same repo by developers who already have write access. Even if the file requires a review by the user, the user is allowed to modify and review those files.
+
+So there is no option to limit the write access to the workflow files.
+
+### Solution 2: Using branch protection rules and environment secrets
+
+You can avoid using repository secrets. If all your secrets are [environment secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-an-environment) the MegaLinter action would not have access to the secrets. Only the jobs linked to the environments could have access to those secrets.
+
+We think this is a good practice anyway because normally secrets are related to one environment. But in some cases, having a repository secret makes sense, so this solution is not possible always.
```
Comment (cgbosse) [inline README.md:98]: ... sense, so this solution is not **""always""** possible.

**Review by cgbosse — COMMENTED** [review]
Some minor improvements in the flow of the text and spelling.

**Comment by hosom** [comment]
:wave:

First of all, thanks for giving us the opportunity to collaborate!

I've included some initial thoughts and feedback below. My team is global and may have additional insight to offer, so I may add more tomorrow as feedback flows in.

I haven't observed any overtly incorrect or false information in this blog post. I think that there's some additional knowledge that is worth sharing though.

> Currently, the [default permissions for all workflows](https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/) in any organization are:

These defaults are configurable within your repository and it's probably worth mentioning that fact. Check out the options under **Settings** | **Actions** | **General** | **Workflow Permissions**. The options here allow you to change the `GITHUB_TOKEN` to `read` by default (a more safe option).

These options also allow you to restrict whether or not Actions can create or approve pull requests.

> ### Solution 1: Using CODEOWNERS configuration
> One of the solutions we thought of was using the [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) configuration.
> 
> If actions can only get secrets from arguments and we are only passing the `GITHUB_TOKEN` to the action, the only way to get the secrets is by modifying a workflow to export the secrets. The CODEOWNERS file allows the repo admins to specify who should review changes on certain files. You could configure a team that has access to the `.\github\workflows` directory but this would not work because:
> 
> * CODEOWNERS only works on pull requests. And in that case, `contents: write` permission is not granted.
> * We are talking about branches created on the same repo by developers who already have write access. Even if the file requires a review by the user, the user is allowed to modify and review those files.
> 
> So there is no option to limit the write access to the workflow files.
> 
> ### Solution 2: Using branch protection rules and environment secrets
> You can avoid using repository secrets. If all your secrets are [environment secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-an-environment) the MegaLinter action would not have access to the secrets. Only the jobs linked to the environments could have access to those secrets.
> 
> We think this is a good practice anyway because normally secrets are related to one environment. But in some cases, having a repository secret makes sense, so this solution is not possible always.

Combining these options is probably the best option. 

You should use a `CODEOWNERS` to require PR reviews of workflow related files. This includes not only the workflows themselves, but also of scripts that you execute from the workflows.

Environments are a great way to isolate secrets. Combine environments with environment rules that dictate which branches that they can be accessed from for the best isolation of your secrets.

As a general rule, I would advise against using `organization` level secrets for anything that you would want to be kept for any amount of time. We aren't going to leak that secret, but it will be quickly accessible by more things than you can count.

Repository secrets can be useful for secrets that are used in nearly every or every job within the repository, but I don't recommend them for anything that has `write` on a system or `read` to sensitive information.

Finally, build multiple environments for your repository and place jobs into environments depending on what access they need.

There are a lot of caveats and subtleties in Actions security, but putting this all together should result in:

* Branch protection rules preventing anyone, including Actions from pushing to `main` or your default branch.
* If you aren't using Actions to open pull requests, consider disabling its ability to do so. 
* A `CODEOWNERS` file preventing Actions from both opening and approving a PR to modify Actions.
* A `GITHUB_TOKEN` with reduced privileges either by changing the defaults at the repository level or explicitly defining your required permissions at the workflow level.
* Environment protections isolating sensitive credentials from general purpose linters and unit test jobs.
* In extreme cases, Environment rules can be used to prevent workflows from executing until approved manually.

**Comment by josecelano** [comment]
Thank you @hosom for your accurate feedback.

### Default token permissions

> These defaults are configurable within your repository and it's probably worth mentioning that fact. Check out the options under Settings | Actions | General | Workflow Permissions. The options here allow you to change the GITHUB_TOKEN to read by default (a more safe option).

Regarding changing default token permissions in your organization I have mentioned it here:

> What we have done is:
>
> 1. Change the default token permissions for the organization. You can [change the default permissions](https://docs.github.com/en/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization#configuring-the-default-github_token-permissions) granted to the GITHUB_TOKEN to read-only.
>
> 2. Remove all permissions on the workflow for the token. You can do it by adding the following to the .github/workflows/mega-linter.yml file: permissions: {}.
>
> 3. Depending on what you want the MegaLinter to do you will need to grant some specific permissions. For example pull-requests: write if you want it to write comments on pull requests or contents: write if you want it to auto-fix your code and push the changes.

but maybe we are mixing too many things in the article. We might split it into 3 articles:

1. Sharing your secrets with third-party actions (the MegaLinter problem)
2. How GitHub Actions can get access to secrets.
3. Best practices or good patterns for configuration.

### Good practices and patterns

Regarding point 3 I think your list is a very good list:

There are a lot of caveats and subtleties in Actions security, but putting this all together should result in:

> - Branch protection rules prevent anyone, including Actions from pushing to main or your default branch.
> - If you aren't using Actions to open pull requests, consider disabling its ability to do so.
> - A CODEOWNERS file preventing Actions from both opening and approving a PR to modify Actions.
> - A GITHUB_TOKEN with reduced privileges either by changing the defaults at the repository level or explicitly defining your required permissions at the workflow level.
Environment protections isolating sensitive credentials from general purpose linters and unit test jobs.
> - In extreme cases, Environment rules can be used to prevent workflows from executing until approved manually.

I would add maybe an example branching strategy that allows implementing this rules. For example, with GitFlow you can run MegaLinter only with the `develop` branch or feature branches but not with the `main` branch. The `main` branch could be linked to the `production` environment, and could contain the secrets. The MegaLinter could be able to push to the `develop` branch if you want to push auto-fixes. I have a question regarding this option. You can enable only certain users to push into a branch:

image

But as far as I know when you run an scheduled workflow and you push, commits are created 
by `github-actions[bot]`. For example:

```s
commit 7dc23dfb585b25f0d155bde47347defaf81e47d0 (HEAD -> main, origin/main, origin/HEAD)
Author:     github-actions[bot] <github-actions[bot]@users.noreply.github.com>
AuthorDate: Thu Jun 2 08:19:03 2022 +0000
Commit:     github-actions[bot] <github-actions[bot]@users.noreply.github.com>
CommitDate: Thu Jun 2 08:19:03 2022 +0000
```

And I suppose you cannot use that account `github-actions[bot]` in the list of allowed users. Therefore I suppose It forces you to create a new GitHub Account for automatic commits and pushes if you want the Bot to have access to the protected branch. It's something we are already doing with our [bot account](https://github.com/nautilus-cyberneering-bot) but I would like to confirm that that would be the only alternative.

### Other questions

- I would like to confirm that actions do not have access to the secrets unless you pass the secrets to them via inputs, arguments, environment vars or input default values in `action.yml`.
- I think I read somewhere else that contexts could be stored on disk and accessed by actions, but I have not found it. I suppose the `secrets` context is passed on the fly to the runner when a new job is started and it's only stored in memory. Of course, there some cases (for example when you use an in-line shell script) where some variable values could be cached on disk. But in general, if you have a workflow with only one action that does nothing, there is no env var accessible by the action or any other way to get a secret from inside the action code.

**Comment by hosom** [comment]
> but maybe we are mixing too many things in the article. We might split it into 3 articles:

I think that would be beneficial. This is a deep and complicated topic to cover in just one.

> The MegaLinter could be able to push to the `develop` branch if you want to push auto-fixes. I have a question regarding this option. You can enable only certain users to push into a branch

I think that if you want MegaLinter to run and commit autofixes, your best option may be to include it as a required check for pull requests to merge into your default branch and then configure it to run on all pull requests. Set the workflow to run with an environment that doesn't contain your secrets and on the `pull_request` event. 

MegaLinter would then commit its proposed changes to the PR and the modified code would be reviewed by a human and then merged.

> And I suppose you cannot use that account github-actions[bot] in the list of allowed users. Therefore I suppose It forces you to create a new GitHub Account for automatic commits and pushes if you want the Bot to have access to the protected branch. It's something we are already doing with our [bot account](https://github.com/nautilus-cyberneering-bot) but I would like to confirm that that would be the only alternative.

That's correct. You'd need a [machine user](https://docs.github.com/en/developers/overview/managing-deploy-keys#machine-users) with its own PAT that you reference as a secret to make this function properly.

> * I would like to confirm that actions do not have access to the secrets unless you pass the secrets to them via inputs, arguments, environment vars or input default values in action.yml.

I am fairly confident the answer to this question is yes. I'll confirm with the Actions development team.

> * I think I read somewhere else that contexts could be stored on disk and accessed by actions, but I have not found it. I suppose the secrets context is passed on the fly to the runner when a new job is started and it's only stored in memory. Of course, there some cases (for example when you use an in-line shell script) where some variable values could be cached on disk. But in general, if you have a workflow with only one action that does nothing, there is no env var accessible by the action or any other way to get a secret from inside the action code.

I'm not sure about this one. I'll confirm with the Actions development team.

**Comment by josecelano** [comment]
hi, @hosom thank you!, that was the kind of wisdom I was looking for. I know that it's difficult to come up with a couple of rules of thumb but at least I think this would be useful to know if you are safe and how to adjust the settings to keep things safe.

At least for me, your answers are very important because it's something I miss in the official documentation.

Regarding question 2, it confirms my impression that the jobs should be the "unit of work" while defining access rules to your secrets. We could rephrase it this way: if you have a job linked to an environment you should act as if all actions in that job could have access to all the environment secrets, or do not mix actions with different levels of access to your secrets.

Regarding using CODEOWNERS, when you said "some workflows can be triggered based on branches" I suppose you mean code review for that change will work only for pull requests and not for direct pushes to that branch. Is that right?

And finally, regarding your last point: "Restricting an environment to only run off of a specific branch is the biggest defence you have.". I think that's a key point in GitHub Actions design. And probably the question you always have to ask when checking the security. Does the action have "write" access to the branch? If yes, then It could have access to the secrets and you should double-check your protection rules.

**Comment by hosom** [comment]
> if you have a job linked to an environment you should act as if all actions in that job could have access to all the environment secrets, or do not mix actions with different levels of access to your secrets.

I think this is a wise assumption.

> Regarding using CODEOWNERS, when you said "some workflows can be triggered based on branches" I suppose you mean code review for that change will work only for pull requests and not for direct pushes to that branch. Is that right?

Yes. CODEOWNERS prevents the code from making its way to the default branch, however, some workflow events can be triggered before the workflow has been merged into the default branch. For example, workflows that trigger off of `pull_request` can be added to a pull request and they will run before they are in the default branch. 

Our documentation kind of explains this in reverse on this page: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows

If you look at each event, it will in many cases say:

> Note: This event will only trigger a workflow run if the workflow file is on the default branch.

If that statement is missing, you can trigger the workflow without first merging it into the default branch. Check out the conditions laid out for pull request to see an example of an event that would allow you to run a workflow WITHOUT a PR review: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

> And finally, regarding your last point: _"Restricting an environment to only run off of a specific branch is the biggest defence you have."_. I think that's a key point in GitHub Actions design. And probably the question you always have to ask when checking the security. Does the action have "write" access to the branch? If yes, then It could have access to the secrets and you should double-check your protection rules.

Correct. Sensitive secrets in the environment and write access to the branch that it can run off of within Actions is a recipe for a bad time.

**Comment by josecelano** [comment]
hi, @cgbosse, @hosom 

I'm closing this PR without merging it because I have moved the content to [another repo](https://github.com/Nautilus-Cyberneering/secure-git-guide) where we are collecting articles like this.

I have split it into two articles:

- [How GitHub Actions can get access to secrets](https://nautilus-cyberneering.github.io/secure-git-guide/013_How-github-actions-can-get-access-to-secrets/)
- [GitHub Actions: sharing your secrets with third-party actions](https://nautilus-cyberneering.github.io/secure-git-guide/014_Sharing-GitHub-secrets-with-third-party-actions/)

@cgbosse thank you very much for your review.

@hosom thank you for your feedback. I have added your ideas and sometimes just copied/pasted your sentences. I hope that's OK for you. I have added your name in the "credits" section of both articles.

I'm going to keep this repo for the source code examples used in the articles.

**Comment by hosom** [comment]
That's fine by me. Thanks for the update.

**Pull request closed without merging**